### PR TITLE
[TEVA-2192] Authentication fallback to use organisation publishers

### PR DESCRIPTION
**Pull request by cesidio**
## Jira ticket URL
https://dfedigital.atlassian.net/browse/TEVA-2192

### CODE REVIEW

**Comment by martin-bangoura** [comment]
Review app deployed to <https://teaching-vacancies-review-pr-3083.london.cloudapps.digital>

File: app/controllers/publishers/sign_in/email/sessions_controller.rb [hunk 1]
```diff
@@ -37,20 +41,19 @@ def redirect_signed_in_publishers
...
-    params_login_key = params[:login_key]
-    begin
-      EmergencyLoginKey.find(params_login_key)
-    rescue StandardError
-      nil
+  def process_login_key
+    login_key = EmergencyLoginKey.find_by(id: params[:login_key])
+    return @reason_for_failing_sign_in = "no_key" unless login_key
+
+    if !login_key.expired?
```
Comment (csutter) [inline app/controllers/publishers/sign_in/email/sessions_controller.rb:52]: Can we flip those and just have an else?

File: app/controllers/publishers/sign_in/email/sessions_controller.rb [hunk 2]
```diff
@@ -2,13 +2,22 @@ class Publishers::SignIn::Email::SessionsController < ApplicationController
...
-  before_action :redirect_for_dsi_authentication, only: %i[new create check_your_email change_organisation choose_organisation]
-  before_action :redirect_unauthorised_publishers, only: %i[create]
+  before_action :redirect_for_dsi_authentication, only: %i[new create check_your_email choose_organisation]
 
   def create
-    session.update(publisher_organisation_id: params[:organisation_id])
-    trigger_sign_in_event(:success, :email)
-    redirect_to organisation_path
+    publisher = Publisher.find(session[:publisher_id])
+    organisation = Organisation.find(params[:organisation_id])
```
Comment (csutter) [inline app/controllers/publishers/sign_in/email/sessions_controller.rb:9]: Can we scope this to the publisher? i.e.:

```ruby
organisation = publisher.organisations.find(params[:organisation_id])
```

**Review by csutter — APPROVED** [review]
I think we need to scope the organisation find to make sure it can't be overridden by an arbitary ID.

Otherwise looks great 👍

File: spec/system/hiring_staff_can_sign_in_by_email_spec.rb [hunk 3]
```diff
@@ -217,7 +217,6 @@
             freeze_time do
               expect { visit auth_email_choose_organisation_path(login_key: login_key.id) }
                 .to have_triggered_event(:publisher_sign_in_attempt)
-                .with_base_data(user_anonymised_publisher_id: anonymised_form_of(publisher.oid))
```
Comment (cpjmcquillan) [inline spec/system/hiring_staff_can_sign_in_by_email_spec.rb:220]: Is this data not there anymore/is it in the data struct instead?

And can we make sure the data is there if it isn't so we keep the performance analysis wheels turning smoothly?

**Review by cpjmcquillan — APPROVED** [review]

**Review by cpjmcquillan — APPROVED** [review]

**Pull request merged**